library/alloc/src/collections/btree/map.rs

```diff
@@ -1775,8 +1781,8 @@ impl<'a, K: 'a, V: 'a> DrainFilterInner<'a, K, V> {
         while let Ok(mut kv) = self.cur_leaf_edge.take()?.next_kv() {
             let (k, v) = kv.kv_mut();
             if pred(k, v) {
-                *self.length -= 1;
-                let (kv, pos) = kv.remove_kv_tracking(|_| self.emptied_internal_root = true);
+                self.length -= 1;
```

Comment:
This seems potentially problematic.

If the length is not updated at each step, and not (e.g.,) set to zero when we create DrainFilter, then if user code `mem::forget`'s a DrainFilter they could observe a BTreeMap with non-zero length despite having zero elements (or any other count, in theory).

I think we'll want to do the same thing Vec::drain_filter does where on construction we set the length of the vector to zero -- https://doc.rust-lang.org/nightly/src/alloc/vec.rs.html#2357-2369.

Can you check the other cases of length being copied instead of referenced directly from the map? (e.g., I suspect OccupiedEntry has the same problem?)

Reply:
Wow, that is a game changer! Then I want to return my #74762 and #75257 and get my money back!

Seriously, I'm puzzled at what level of consistency we want. Do we just want the drop handler of the map to survive? Or a fully functional map, with fewer or no elements? I suppose leaks are allowed? I'll be hunting down the history of those 3 vec lines.

Reply:
The basic level of consistency we require is "no memory unsafeties ever, for any possible safe client".

If achieving that requires leaking elements, so be it.

Reply:
For the history of the `Vec` code, see http://cglab.ca/~abeinges/blah/everyone-poops/.

Reply:
So it is in The Book (and  leak does not even appear in the index of my O'Reilly book). I'm not 100sure that operations as searching a tree with an empty root is undefined behavior but it's pretty likely then that #74762 made drain_filter unsafe.

Reply:
Hm okay, if necessary we can revert that PR -- I may have missed that -- my impression from back then I think was that skipping the destructor wouldn't be a problem as the length would be zero still (as we tracked the length via `&mut self.length` vs. copying it. But I definitely may have missed something!

Reply:
Would be good to get a test case for this (and then Miri will also tell us whether we truly have UB).

Reply:
I see 3 ways about this:
- As Mark said, zero out the length, and make sure that all operations on zero-length maps stay away from the root
- Also `take` the root, put it back in the drop handler
- Revert #74762, assuming there really isn't UB there (I've been trying to prove it was close, but Miri remains silent), and it should be possible to sail around UB in this PR here too.

Reply:
taking the root seems like the safest thing, right? We should probably take the root and zero the length though.

We can be less leak-happy eventually, if we want, but I'm personally in favor of not worrying too much about leaking if someone leaks drain filter.

Reply:
> taking the root seems like the safest thing, right?

I think it's definitely better than the first option, and it's the only way to make drain_filter more efficient in the long term.

> We should probably take the root and zero the length though.

Yes, and I did not mean to imply we "also" have to check operations on zero-length maps, the map simply temporarily reverts to its newborn state.

Reply:
> Hm okay, if necessary we can revert that PR -- I may have missed that

That PR didn't involve length but only postponed the `pop_internal_level`. I've not written the test yet, but if you skip that `pop_internal_level`, `search` will most likely be fine, but `range_search` should not appreciate the empty root with equal left and right edges.

But it's peculiar use of an unstable API, so not something to urgently address I think? Compared to say the UB in #73971.

Reply:
> whether we truly have UB

Oh, I am aiming for something more spectacular than that. You've given me a new weapon, it's too hot, some software has to suffer.

Reply:
In my book there us nothing more spectacular than UB, but now you made me curious. ;)

Reply:
Even unstable safe APIs shouldn't cause memory unsafety if misused; if pop_internal_level must be called to avoid unsafety then we should be doing so.

https://github.com/rust-lang/rust/pull/73971 is, IIRC, only a hypothetical in that it's a violation of stacked borrows which is both not describing Rust's memory model today (just a possible memory model). We should definitely fix it, as stacked borrows is our current most worked on proposal for a memory model, but to my knowledge the failure miri is catching there is not something that can actually cause unsoundness on stable Rust today in some sense.

Reply:
I'm pretty sure there is no problem whatsoever with skipping the drop handler... because it did nothing.

So trees with empty internal nodes have existed for a while, and none of the problems I thought there were, and that I tested, actually registered any seismic activity. Also, the UB I thought there was in #74762 was worse than I realized, which makes it more sure it wasn't UB to begin with.

The "potentially problematic" use of `mem:forget` within this PR obviously results in access violations - if length is too high, `IntoIter` runs off the edge of the tree into the abyss.

> `take` the root, put it back in the drop handler

Easier said than done. It's borrowed by the leaf edge cursor, and I don't think you can pass around a borrowee by value in a `DrainFilter` instance.